### ISSUE TITLE
Differential privacy for training

### ISSUE THREAD

**Issue by aymuos15** [aymuos15/AutoMONAI#30]
## Motivation

Medical imaging models trained on patient data can memorize and leak sensitive information. Differential privacy (DP) provides formal guarantees that individual training samples cannot be identified from the trained model. This is increasingly required for regulatory compliance and is a natural complement to federated learning (#28).

### Approach: DP-SGD via Opacus

[Opacus](https://opacus.ai/) (by Meta) is the standard library for training PyTorch models with differential privacy. It wraps the optimizer to clip per-sample gradients and add calibrated noise.

### CLI flags
```
--dp                    Enable differential privacy (default: false)
--dp_epsilon            Target privacy budget ε (default: 10.0)
--dp_delta              Target δ, typically 1/N (default: auto — set to 1/dataset_size)
--dp_max_grad_norm      Per-sample gradient clipping norm (default: 1.0)
--dp_noise_multiplier   Noise multiplier σ (default: auto — computed from ε, δ, epochs)
```

Users set either `--dp_epsilon` (and let noise be auto-computed) or `--dp_noise_multiplier` (and get the resulting ε reported).

### Core implementation

**`automonai/core/dp.py`**:
```python
def make_private(model, optimizer, train_loader, target_epsilon, delta, epochs, max_grad_norm):
    """Wrap model/optimizer/loader with Opacus DP guarantees.
    
    Returns (dp_model, dp_optimizer, dp_loader, privacy_engine).
    """
```

**`run.py` changes**:
- When `--dp` is set, call `make_private()` after model/optimizer creation but before Fabric setup
- Log privacy spent (ε consumed) per epoch to W&B
- Print final privacy budget at training end
- Opacus requires `batch_first=True` and doesn't support all layer types — validate model compatibility upfront and raise a clear error for incompatible architectures

### Model compatibility
Opacus requires replacing certain layers:
- `BatchNorm` → `GroupNorm` (Opacus does this automatically via `ModuleValidator`)
- Some custom MONAI layers may need manual fixes

Add a validation step:
```python
from opacus.validators import ModuleValidator
errors = ModuleValidator.validate(model, strict=False)
if errors:
    model = ModuleValidator.fix(model)  # auto-fix where possible
```

### W&B logging
- `dp/epsilon_spent` — cumulative ε after each epoch
- `dp/delta` — fixed δ value
- `dp/noise_multiplier` — σ used
- `dp/max_grad_norm` — clipping norm
- Final `dp/total_epsilon` in run summary

### UI changes
- **Differential Privacy** toggle in Generate tab (advanced section)
- When enabled: epsilon input, max grad norm input
- Warning banner: "DP training is slower and may reduce model accuracy"
- Configs tab: show 🔒 badge on DP-enabled configs

### Interaction with federated learning (#28)
- DP + FL = each client trains with DP-SGD locally before sending updates
- Flower supports this via its DP modules — compose with `AutoMONAIClient`
- Combined privacy accounting across rounds

### Dependencies
```
pip install opacus
```

## Privacy budget guidance (in docs)
| ε | Interpretation |
|---|---|
| 1–3 | Strong privacy, significant accuracy cost |
| 5–10 | Moderate privacy, practical for most tasks |
| 10–50 | Weak formal guarantee, minimal accuracy impact |

## Out of scope
- Secure aggregation / homomorphic encryption (orthogonal to DP)
- PATE (Private Aggregation of Teacher Ensembles) — different approach, future issue
- Per-class privacy budgets
